poly-commitment/src/lagrange_cache.rs

```diff
@@ -0,0 +1,201 @@
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+    ) -> Option<Vec<PolyComm<G>>>;
+
+    fn cache_lagrange_basis(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+        basis: &Vec<PolyComm<G>>,
+    );
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct FileCache<G> {
+    cache_dir: PathBuf,
+    point_type: PhantomData<G>,
+}
+
+impl<G> FileCache<G> {
+    pub fn new(cache_dir: PathBuf) -> Self {
+        if !cache_dir.exists() {
+            fs::create_dir_all(&cache_dir).unwrap();
+        }
+        FileCache {
+            cache_dir,
+            point_type: PhantomData,
+        }
+    }
+}
+
+impl<G: CommitmentCurve> LagrangeCacheTrait<G> for FileCache<G> {
+    type CacheKey = PathBuf;
+
+    fn lagrange_basis_cache_key(&self, g: &Vec<G>, domain: &D<G::ScalarField>) -> Self::CacheKey {
+        let mut hasher = DefaultHasher::new();
+        g.hash(&mut hasher);
+        domain.size.hash(&mut hasher);
+        self.cache_dir
+            .clone()
+            .join(format!("lagrange_basis_{:}", hasher.finish()))
+    }
+
+    fn load_lagrange_basis_from_cache(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+    ) -> Option<Vec<PolyComm<G>>> {
+        let cache_key = self.lagrange_basis_cache_key(g, domain);
+        if Path::exists(&cache_key) {
+            let f = File::open(cache_key.clone()).expect(&format!(
+                "Missing lagrange basis cache file {:?}",
+                cache_key
+            ));
+            let basis = rmp_serde::decode::from_read(f).expect(&format!(
+                "Error decoding lagrange cache file {:?}",
+                cache_key
+            ));
+            println!(
+                "Successully loaded lagrange basis from cache {:?}",
+                cache_key
+            );
+            Some(basis)
+        } else {
+            None
+        }
+    }
+
+    fn cache_lagrange_basis(
+        &self,
+        g: &Vec<G>,
+        domain: &D<G::ScalarField>,
+        basis: &Vec<PolyComm<G>>,
+    ) {
+        let cache_key = self.lagrange_basis_cache_key(g, domain);
+        if Path::exists(&cache_key) {
+            println!("Lagrange basis cache file {:?} already exists", cache_key);
+            return;
+        } else {
+            println!("Caching lagrange basis to file {:?}", cache_key);
+            let mut f = File::create(cache_key.clone()).expect(&format!(
+                "Error creating lagrabnge basis cache file {:?}",
+                cache_key
+            ));
+            rmp_serde::encode::write(&mut f, basis).expect(&format!(
+                "Error encoding lagrange basis to file {:?}",
+                cache_key
+            ));
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GoogleCloudCache<G> {
+    url: String,
+    point_type: PhantomData<G>,
+}
+
+impl<G> GoogleCloudCache<G> {
+    pub fn new(url: PathBuf) -> Self {
+        todo!("GoogleCloudCache::new")
+    }
+}
```

Comment:
In theory we can use any remote file store not a google cloud bucket. I think something like `RemoteCache` is a slightly better naming scheme.

Reply:
Now that I think about it, I see why this makes sense. Every struct should have a different fetching implementation, so you would implement this separately per service.